Add support for additional audio formats

septante opened the pull request:
Resolves #707

mrDIMAS commented:
IMO the all submodules of `decoder` module should be deleted and internals of `impl Decoder` should be replaced with `symphonia`. It does not really worth it to replace decoders one by one.

septante commented:
Added a generic implementation and removed the submodules. Still need to add an error type, but it seems like the tests pass and the play_sound example mentioned in the issue still runs.

mrDIMAS commented:
I think it can be done in this PR.

septante commented:
Should I change the callers to return Errors as well, or try to keep the original signatures that return DataSource as the Err type?

septante commented:
Changed the signatures to Errors for now and fixed all the warnings/errors. The dependency on hound for wav files can't be removed for now, since it is used in the write_wav example. Not sure if the example should be rewritten or just left as is.

mrDIMAS commented on fyrox-sound/src/decoder/mod.rs:
```diff
@@ -18,84 +18,172 @@
+
+        let mut reader = res.format;
+        let tracks = reader.tracks();
+        let first_track = tracks.first().unwrap();
```
Replace `unwrap` call with error mapping (`ok_or_else`) and use `?` with it.

septante commented on fyrox-sound/src/decoder/mod.rs:
```diff
@@ -18,84 +18,172 @@
+                vec = samples.to_vec();
+            }
+        }
+        let samples = vec.into_iter();
```
Done

septante commented on fyrox-sound/src/decoder/mod.rs:
```diff
@@ -18,84 +18,172 @@
+            )
+            .is_err()
+        {
+            println!("Failed to seek vorbis/ogg?")
```
Should this function return the error instead? It seems a little odd that rewind does but seek doesn't.

mrDIMAS replied:
Yeah, you're right.

septante replied:
Done

septante commented on fyrox-sound/src/decoder/mod.rs:
```diff
@@ -18,84 +18,172 @@
+            if let Ok(packet) = self.reader.next_packet() {
+                if let Ok(decoded) = self.decoder.decode(&packet) {
+                    let buffer: AudioBuffer<Self::Item> = decoded.make_equivalent();
+                    let samples = buffer.chan(0);
```
It seems like
```rust
let left_samples = buffer
    .chan(Channels::FRONT_LEFT.bits() as usize)
    .into_iter();
let right_samples = buffer
    .chan(Channels::FRONT_RIGHT.bits() as usize)
    .into_iter();
```
and then interleaving should work? Either with the itertools crate or a hand-written implementation.

septante replied:
Would iterating over the channels and interleaving all of the slices be a reasonable approach?

mrDIMAS replied:
It may work, but it probably will be relatively slow to read than from an interleaved array.

septante replied:
Implemented stereo audio, though when trying to play the file found [here](https://www.aoakley.com/articles/2018-08-24-stereo-test.php) I noticed that it only plays the first "front center", when the entire clip includes side left/right, followed by repeating each of the three one more time. Not sure if this is a limitation of symphonia or if there's something else I'm missing.

septante replied:
Never mind, the thread wasn't sleeping for long enough. I think that's everything done for this PR?

mrDIMAS commented on fyrox-sound/src/decoder/mod.rs:
```diff
@@ -131,11 +131,30 @@ impl Decoder {
     }
 
+    fn interleaved(buffer: &mut AudioBuffer<f32>) -> Vec<f32> {
+        let channels = buffer.spec().channels.count();
```
Naming is confusing here. This should be `channel_count`.

septante replied:
Done

mrDIMAS commented on fyrox-sound/src/decoder/mod.rs:
```diff
@@ -131,11 +131,30 @@ impl Decoder {
+        let channels = buffer.spec().channels.count();
+        let frames = buffer.frames();
+
+        let mut samples = Vec::with_capacity(channels);
```
This should be `channels`

septante replied:
Done

mrDIMAS commented:
Could you please check if all examples from `fyrox-sound` are working (especially the streaming one)? If everything is ok, I'll merge the PR right away.

septante commented:
Hm, streaming appears to be broken. I'll take a look

septante commented:
Some notes so I don't forget: the cause of the panic seems to be an unwrap occurring [here](https://docs.rs/symphonia-format-ogg/0.5.4/src/symphonia_format_ogg/demuxer.rs.html#124), because the check [here](https://docs.rs/symphonia-format-ogg/0.5.4/src/symphonia_format_ogg/demuxer.rs.html#308) returns None. The solution should just be to change the byte_len implementation in buffer/mod.rs.

septante commented:
Fixed! I wasn't sure if I could use the length of the vector as the byte_len value for DataSouce::Memory, so I just used None for now. I left a TODO note in the last commit mentioning it, if you want to take a look.

mrDIMAS commented on fyrox-sound/src/buffer/mod.rs:
```diff
@@ -181,15 +187,63 @@ impl Seek for DataSource {
+    fn byte_len(&self) -> Option<u64> {
+        match self {
+            DataSource::File { path, data: _ } => {
+                if let Ok(f) = File::open(path) {
```
Don't use `File::open` here, it is not compatible with neither Android nor Wasm. Instead, use `MediaSource::byte_len(data)` itself, it should work since data is `Box<dyn FileReader>` which implements all the traits for `MediaSource` trait.

septante replied:
I'm not seeing an implementation for MediaSource on FileReader? Trying to use MediaSource::byte_len(data) fails to compile.

mrDIMAS replied:
You can always do this 

```rust
let byte_len = data.seek(SeekFrom::End(0))?;
data.seek(SeekFrom::Start(0))?;
```

septante replied:
Hm, that doesn't seem to work either, since self is not a mut reference.

mrDIMAS replied:
Ok, just add a `byte_len` method to `FileReader` and implement the trait for `std::fs::File` and remove auto-impl for it. The impl could be this https://docs.rs/symphonia-core/latest/src/symphonia_core/io/mod.rs.html#69

septante replied:
Should I add a default implementation that returns None to avoid breaking other uses of FileReader?

septante replied:
Oh, it still needs to be implemented for the other types regardless.

mrDIMAS commented on fyrox-sound/src/buffer/mod.rs:
```diff
@@ -181,15 +187,63 @@ impl Seek for DataSource {
+                    None
+                }
+            }
+            // TODO: can we use the length of the underlying vector here?
```
Yeah, you can use the length here.

septante replied:
Fixed
